README.md
# psbcSpeedUp

[![CRAN](http://www.r-pkg.org/badges/version/psbcSpeedUp)](https://cran.r-project.org/package=psbcSpeedUp)


This is a C++ speed-up and extended version of the R-pakcage [psbcGroup](https://CRAN.R-project.org/package=psbcGroup).
It implements the Bayesian Lasso Cox model ([Lee et al., 2011](https://doi.org/10.2202/1557-4679.1301)) and the Bayesian Lasso Cox with mandatory variables ([Zucknick et al., 2015](https://doi.org/10.1002/bimj.201400160)).
Bayesian Lasso Cox models with other shrinkage and group priors ([Lee et al., 2015](https://doi.org/10.1002/sam.11266)) are to be implemented later on.

## Installation

Install the latest released version from [CRAN](https://CRAN.R-project.org/package=psbcSpeedUp)

```r
install.packages("psbcSpeedUp")
```

Install the latest development version from GitHub

```r
#install.packages("remotes")
remotes::install_github("ocbe-uio/psbcSpeedUp")
```

## Examples

### Run a Bayesian Lasso Cox with mandatory variables

Data set `exampleData` consists of six components:
survival times `t`,
event status `di`,
covariates `x`,
number of genomics variables `p`,
number of clinical variables `q` and
true effects of covariates `beta_true`.
See `?exampleData` for more information of the data.

To run a Bayesian Lasso Cox model for variable selection of the first $p$ genomics variables and inclusion of $q$ mandatory variables, one can specify arguments of the main function `psbcSpeedUp()` with `p = p` and `q = q`.
If the arguments `p` and `q` are unspecified, the Bayesian Lasso Cox model does variable selection for all covariates, i.e., by default `p = ncol(survObj$x)` and `q = 0`.

```r
# Load the example dataset
data("exampleData", package = "psbcSpeedUp")
p <- exampleData$p
q <- exampleData$q
survObj <- exampleData[1:3]

# Set hyperparameters (see help file for specifying more hyperparameters)
mypriorPara <- list('eta0'=0.02, 'kappa0'=1, 'c0'=2, 'r'=10/9, 'delta'=1e-05,
'lambdaSq'=1, 'sigmaSq'= runif(1, 0.1, 10), 'beta.prop.var'=1, 'beta.clin.var'=1)

# run Bayesian Lasso Cox
library("psbcSpeedUp")
set.seed(123)
fitBayesCox <- psbcSpeedUp(survObj, p=p, q=q, hyperpar=mypriorPara,
nIter=1000, burnin=500, outFilePath="/tmp")
```

```
Running MCMC iterations ...
[##################################################] 100%
DONE, exiting!
```

### Plot posterior estimates of regression cofficients

The function `psbcSpeedUp::plot()` can show the posterior mean and 95% credible intervals of regression coefficients.

```r
plot(fitBayesCox)
```

<img src="man/figures/README_plot_beta.png" width="50%" />

## References

> Kyu Ha Lee, Sounak Chakraborty, Jianguo Sun (2011).
> Bayesian variable selection in semiparametric proportional hazards model for high dimensional survival data.
> _The International Journal of Biostatistics_, 7:1. DOI: [10.2202/1557-4679.1301](https://doi.org/10.2202/1557-4679.1301).

> Kyu Ha Lee, Sounak Chakraborty, Jianguo Sun (2015).
> Survival prediction and variable selection with simultaneous shrinkage and grouping priors.
> _Statistical Analysis and Data Mining_, 8:114-127. DOI:[10.1002/sam.11266](https://doi.org/10.1002/sam.11266).

> Manuela Zucknick, Maral Saadati, Axel Benner (2015).
> Nonidentical twins: Comparison of frequentist and Bayesian lasso for Cox models.
> _Biometrical Journal_, 57:959-981. DOI:[10.1002/bimj.201400160](https://doi.org/10.1002/bimj.201400160).


NEWS.md
# TODO in next versions

* use Bandicoot: Armadillo C++ library for GPU
* remove Rcpp::List object, instead writing all output objects into files
* allow multiple chains and omp for parallelisation
* add R functions for feature (stability) selection
* add R functions for survival predictions
* extend C++ source code for implementing other shrinkage and group priors

# psbcSpeedUp 2.0.4

* Improved C++ code to speedup 10-20%
* Declared functions inside PSBC class
* Added configuration files for checking omp, compilers etc.

# psbcSpeedUp 2.0.3

* First CRAN version
* Added adaptve jumping rule
* Fixed bug for q=0 to allowed variable selection for all covariates by default

# psbcSpeedUp 2.0.2

* Optimised the partition of the time axis for gamma process prior

# psbcSpeedUp 2.0.1

* Set armadillio random values and RNG

# psbcSpeedUp 2.0

* First GitHub commit version


src/drive.cpp
#include "drive.h"

#ifndef CCODE
    using Rcpp::Rcout;
    using Rcpp::Rcerr;
#else
    #define Rcout std::cout
    #define Rcerr std::cerr
#endif

#ifdef _OPENMP
extern omp_lock_t RNGlock; /*defined in global.h*/
#endif
/*extern std::vector<std::mt19937_64> rng;*/

// not yet set up openmp
#include <Rcpp.h>
// [[Rcpp::plugins(openmp)]]

using Utils::Chain_Data;

// main function
// (i) import data and parameters; (ii) MCMC algorithm; (iii) export estimates
Rcpp::List drive( const std::string& dataFile, const unsigned int p, const unsigned int q, const std::string& hyperParFile, const std::string& outFilePath,
                 const arma::vec ini_beta, const arma::vec ini_tauSq, const arma::vec ini_h, const arma::uvec groupInd,
                 const unsigned int nIter, const unsigned int nChains, const unsigned int thin, bool rw)
{
    
    // set random seed
    #ifdef _OPENMP
        // NOT yet/easily implement for updateRP().
        // Through a simple test, '#pragma omp parallel for' slowed down the computation of other small functions
        omp_set_nested( 0 );
        omp_set_num_threads( 1 );
        omp_init_lock(&RNGlock);  // init RNG lock for the parallel part
        //arma::arma_rng::set_seed(123);
    #endif
    
    // ###########################################################
    // ## Read Arguments and Data
    // ###########################################################
    
    //Rcout << std::fixed << std::setprecision(11);
    
    Chain_Data chainData; // this initialises the pointers and the strings to ""
    chainData.outFilePath = outFilePath;
    
    PSBC myPSBC;
    
    // read Data and format into usables
    //Rcout << "Reading input files ... " <<  "\n";
    
    Utils::formatData(dataFile, chainData.survData );
    Utils::readHyperPar(hyperParFile, chainData );
    
    // declare all the data-related variables
    double eta0 = chainData.eta0;
    double kappa0 = chainData.kappa0;
    double c0 = chainData.c0;
    double r = chainData.r;
    double delta = chainData.delta;
    
    arma::vec dataTime = chainData.survData.data->col( 0 );
    arma::vec dataDi = chainData.survData.data->col( 1 );
    arma::vec s0 = {0.}; //event times that are actually observed
    s0(0) = 2. * max( dataTime ) - max( dataTime.elem( arma::find(dataTime != max(dataTime)) ) );
    arma::vec s = arma::join_cols( arma::sort( arma::unique( dataTime.elem( arma::find(dataDi==1.) ) ) ), s0 );
    unsigned int J = s.n_elem;
    arma::uvec groupNo = arma::unique( groupInd );
    unsigned int K = groupNo.n_elem;
    
    ind_r_d = ind_r = ind_d = arma::zeros<arma::mat>( chainData.survData.data->n_rows, J );
    d = arma::sum( ind_d.t(), 1 );
    myPSBC.settingInterval_cpp( dataTime, dataDi, s, J, ind_d, ind_r, ind_r_d, d );
    
    arma::vec beta_prop_me;
    be = beta_prop_me = ini_beta;
    arma::mat beta_p = arma::zeros<arma::mat>( (int)(nIter/thin) + 1, p + q );
    beta_p.row( 0 ) = ini_beta.t();
    
    double lambdaSq;
    lambdaSq = chainData.lambdaSq;
    double sigmaSq;
    sigmaSq = chainData.sigmaSq;
    arma::vec tauSq = ini_tauSq;
    h = ini_h;
    
  // tausq: only for genomic variables
    arma::vec tauSq_exp = arma::zeros<arma::vec>( p );
    for( unsigned int i=0; i<K; ++i )
    {
       // groupNoLocation = arma::find( groupInd == groupNo(i) );
        tauSq_exp.elem( arma::find( groupInd == groupNo(i) ) ).fill( tauSq(i) );
    }
   
    double beta_prop_sd = sqrt(chainData.beta_prop_var);
    double beta_clin_sd = sqrt(chainData.beta_clin_var);
    arma::vec sd_bePart2 = arma::zeros<arma::vec>( q );
    sd_bePart2.fill( beta_clin_sd );
    arma::vec sd_be = arma::join_cols( sqrt(sigmaSq * tauSq_exp), sd_bePart2 );
    
    //arma::uvec xId = arma::linspace<arma::uvec>(2, 1+p+q, p+q)
    arma::mat x = chainData.survData.data->cols( arma::linspace<arma::uvec>(2, 1+p+q, p+q) );
    arma::vec xbeta = x * ini_beta;
    
    arma::vec be_normSq = arma::zeros<arma::vec>( K );
    for( unsigned int i=0; i<K; ++i )
        be_normSq(i) = arma::accu( ini_beta.elem( arma::find(groupInd == groupNo(i)) ) % ini_beta.elem( arma::find(groupInd == groupNo(i)) ) );
    
    if( !any( be_normSq ) )
        be_normSq(0) = 0.1;
    
    arma::vec alpha0 = arma::zeros<arma::vec>( 1 + J );
    //H_star(j) = eta0 * arma::pow( s, kappa0 );
    alpha0.subvec( 1, J ) = c0 * eta0 * arma::pow( s, kappa0 );
    arma::vec hPriorSh = arma::diff( alpha0 );
    
    // save mcmc intermediate estimates
    arma::mat h_p = arma::zeros<arma::mat>( (int)(nIter/thin) + 1, h.n_elem );
    h_p.row( 0 ) = h.t();
    arma::mat tauSq_p = arma::zeros<arma::mat>( (int)(nIter/thin) + 1, tauSq.n_elem );
    tauSq_p.row( 0 ) = tauSq.t();
    
    arma::vec sigmaSq_p = arma::zeros<arma::vec>( (int)(nIter/thin) + 1 );
    arma::vec lambdaSq_p = sigmaSq_p;
    sampleRPc_accept = arma::zeros<arma::uvec>( q );
    sampleRPg_accept = arma::zeros<arma::uvec>( p );
    
    
    // ###########################################################
    // ## MCMC loop
    // ###########################################################
    
    //unsigned int tick = 100; // how many iter for each print?
    unsigned int j = 0; // count thinned results
    const int cTotalLength = 50;
    Rcout << "Running MCMC iterations ...\n";
    
    for( unsigned int M=0; M<nIter; ++M )
    {
        
        if( M % 10 == 0 || M == (nIter - 1) )
            Rcout << "\r[" <<   //'\r' aka carriage return should move printer's cursor back at the beginning of the current line
            std::string(cTotalLength * (M+1.) / nIter, '#') <<        //printing filled part
            std::string(cTotalLength * (1. - (M+1.) / nIter), '-') <<  //printing empty part
            "] " << int( (M+1.) / nIter * 100.0 ) << "%\r"; //printing percentage

        //Updating regression coefficients and hyperparameters
        
        if( q > 0 )
            myPSBC.updateRP_clinical_cpp( p, q, x, ind_r, ind_d, ind_r_d, J, beta_prop_me, beta_prop_sd, xbeta, be, h, sd_be, sampleRPc_accept );
        
        if( rw )
        {
            myPSBC.updateRP_genomic_rw_cpp( p, x, ind_r, ind_d, ind_r_d, J, beta_prop_me, beta_prop_sd, xbeta, be, h, sd_be, sampleRPg_accept );
        } else {
            myPSBC.updateRP_genomic_cpp( p, x, ind_r, ind_d, ind_r_d, J, xbeta, be, h, sd_be, sampleRPg_accept );
        }
        
//        if( q > 0 )
//        {
            for( unsigned int i=0; i<K; ++i )
                be_normSq(i) = arma::accu( be.elem( arma::find(groupInd == groupNo(i)) ) % be.elem( arma::find(groupInd == groupNo(i)) ) );
//        } else {
//            be_normSq = be % be;
//        }
        
        h = myPSBC.updateBH_cpp( ind_r_d, hPriorSh, d, c0, J, xbeta );
            
        tauSq = myPSBC.updateTau_GL_cpp( lambdaSq, sigmaSq, be_normSq );

        sigmaSq = myPSBC.updateSigma_GL_cpp( p, be_normSq, tauSq);
        lambdaSq = myPSBC.updateLambda_GL_cpp( p, K, r, delta, tauSq);
        
//        if( q > 0 )
//        {
            for( unsigned int i=0; i<K; ++i )
                tauSq_exp.elem( arma::find(groupInd == groupNo(i)) ).fill( tauSq(i) );
            
            sd_bePart2.fill( beta_clin_sd );
            sd_be = arma::join_cols( sqrt(sigmaSq * tauSq_exp), sd_bePart2 );
//        } else {
//            tauSq_exp = tauSq;
//            sd_be = sqrt(sigmaSq * tauSq_exp);
//        }

        // Save all results
        if( M % thin == 0 )
        {
            ++j;
            beta_p.row( j )  = be.t();
            h_p.row( j )  = h.t();
            tauSq_p.row( j )  = tauSq.t();
            sigmaSq_p( j ) = sigmaSq;
            lambdaSq_p( j ) = lambdaSq;
        }
        
        // adaptve jumping rule
        if( j > 20 )
            for( unsigned int jj=0; jj<p+q; ++jj )
                if( ini_beta( jj ) == beta_p( j - 20, jj ) )
                    beta_prop_me( jj ) = be( jj );
        
    }
    
    arma::uvec accept_rate = join_cols(sampleRPg_accept, sampleRPc_accept);// / (double)(nIter);
    
    // Exit
    Rcout << "\nDONE, exiting! \n";
    
    return Rcpp::List::create(Rcpp::Named("beta.p") =  beta_p,
                              Rcpp::Named("h.p") = h_p,
                              Rcpp::Named("tauSq.p") = tauSq_p,
                              Rcpp::Named("sigmaSq.p") = sigmaSq_p,
                              Rcpp::Named("lambdaSq.p") = lambdaSq_p,
                              Rcpp::Named("accept.rate") = accept_rate
                              );
    
}

// [[Rcpp::export]]
Rcpp::List psbcSpeedUp_internal( const std::string& dataFile, const unsigned int p, const unsigned int q, const std::string& hyperParFile, const std::string& outFilePath,
                                const arma::vec ini_beta, const arma::vec ini_tauSq, const arma::vec ini_h, const arma::uvec groupInd, const unsigned int nIter, const unsigned int nChains, const unsigned int thin, bool rw )
{
  //int status {1};
    Rcpp::List beta_mcmc;
    try
    {
        // status = drive(...);
        beta_mcmc =  drive( dataFile, p, q, hyperParFile, outFilePath, ini_beta, ini_tauSq, ini_h, groupInd, nIter, nChains, thin, rw );
    }
    catch( const std::exception& e )
    {
        Rcout << e.what() << '\n'; // we can use Rcerr here because we're reaching here from R for sure
    }
    
    // return status;
    return beta_mcmc;
}


R/coef.psbcSpeedUp.R
#' @title coef method for class \code{psbcSpeedUp}
#'
#' @description
#' Extract the point estimates of the regression coefficients
#'
#' @name coef.psbcSpeedUp
#'
#' @param object an object of class \code{psbcSpeedUp}
#' @param type type of point estimates of regressions. One of
#' \code{c("mean", "median")}. Default is \code{mean}
#' @param ... not used
#'
#' @return Estimated coefficients are from an object of class \code{psbcSpeedUp}.
#' If the \code{psbcSpeedUp} specified data standardization, the fitted values
#' are base based on standardized data.
#'
#' @examples
#'
#' # Load the example dataset
#' data("exampleData", package = "psbcSpeedUp")
#' p <- exampleData$p
#' q <- exampleData$q
#' survObj <- exampleData[1:3]
#'
#' # Set hyperparameters
#' mypriorPara <- list(
#'   "groupInd" = 1:p, "eta0" = 0.02, "kappa0" = 1, "c0" = 2, "r" = 10 / 9, 
#'   "delta" = 1e-05, "lambdaSq" = 1, "sigmaSq" = runif(1, 0.1, 10), 
#'   "beta.prop.var" = 1, "beta.clin.var" = 1)
#'
#'\donttest{
#' # run Bayesian Lasso Cox
#' library("psbcSpeedUp")
#' set.seed(123)
#' fitBayesCox <- psbcSpeedUp(survObj,
#'   p = p, q = q, hyperpar = mypriorPara,
#'   nIter = 10, burnin = 0, outFilePath = tempdir()
#' )
#' coef(fitBayesCox)
#' }
#'
#' @export
coef.psbcSpeedUp <- function(object, type = "mean", ...) {
  if (!inherits(object, "psbcSpeedUp")) {
    stop("Use only with \"psbcSpeedUp\" objects")
  }

  if (length(type) == 1) {
    if (!type %in% c("mean", "median")) {
      stop("'type' should be one of c('mean', 'median')!")
    }
  } else {
    stop("'type' should be one of c('mean', 'median')!")
  }

  beta_p <- object$output$beta.p[-(1:(object$input$burnin + 1)), ]
  beta_est <- apply(beta_p, 2, type)

  return(beta_est)
}


R/plot.psbcSpeedUp.R
#' @title create a plot of estimated coefficients
#' @description
#' Plot point estimates of regression coefficients and 95\% credible intervals
#'
#' @name plot.psbcSpeedUp
#'
#' @importFrom GGally ggcoef
#' @importFrom ggplot2 xlab ylab
#' @importFrom stats quantile
#'
#' @param x an object of class \code{psbcSpeedUp} or a matrix. If \code{x}
#' is a matrix, use \code{psbcSpeedUp:::plot.psbcSpeedUp(x)}
#' @param type type of point estimates of regression coefficients. One of
#' \code{c("mean", "median")}. Default is \code{mean}
#' @param interval logical argument to show 95\% credible intervals. Default
#' is \code{TRUE}
#' @param ... additional arguments sent to \code{ggplot2::geom_point()}
#'
#' @return ggplot object
#'
#' @examples
#'
#' # Load the example dataset
#' data("exampleData", package = "psbcSpeedUp")
#' p <- exampleData$p
#' q <- exampleData$q
#' survObj <- exampleData[1:3]
#'
#' # Set hyperparameters
#' mypriorPara <- list(
#'   "groupInd" = 1:p, "eta0" = 0.02, "kappa0" = 1, "c0" = 2, "r" = 10 / 9, 
#'   "delta" = 1e-05, "lambdaSq" = 1, "sigmaSq" = runif(1, 0.1, 10), 
#'   "beta.prop.var" = 1, "beta.clin.var" = 1)
#'
#'\donttest{
#' # run Bayesian Lasso Cox
#' library("psbcSpeedUp")
#' set.seed(123)
#' fitBayesCox <- psbcSpeedUp(survObj,
#'   p = p, q = q, hyperpar = mypriorPara,
#'   nIter = 10, burnin = 0, outFilePath = tempdir()
#' )
#' plot(fitBayesCox, color = "blue")
#' }
#'
#' @export
plot.psbcSpeedUp <- function(x, type = "mean", interval = TRUE, ...) {
  if (!(inherits(x, "psbcSpeedUp") | is.matrix(x))) {
    stop("Use only with 'psbcSpeedUp' object or a matrix!")
  }

  if (length(type) == 1) {
    if (!type %in% c("mean", "median")) {
      stop("'type' should be one of c('mean', 'median')!")
    }
  } else {
    stop("'type' should be one of c('mean', 'median')!")
  }

  if (!is.logical(interval)) {
    stop("Argument 'interval' must be a logical value!")
  }

  if (inherits(x, "psbcSpeedUp")) {
    if (is.null(colnames(x$output$beta.p))) {
      x_names <- paste0("x", 1:ncol(x$output$beta.p))
    } else {
      x_names <- colnames(x$output$beta.p)
    }
    beta_p <- x$output$beta.p[-(1:(x$input$burnin / x$input$thin + 1)), ]
  } else {
    if (is.null(colnames(x))) {
      x_names <- paste0("x", 1:ncol(x))
    } else {
      x_names <- colnames(x)
    }
    beta_p <- x
  }

  # pdf("psbcBeta.pdf", height = 5, width = 3.5)
  beta_est <- apply(beta_p, 2, type)
  beta_L <- apply(beta_p, 2, quantile, 0.025)
  beta_U <- apply(beta_p, 2, quantile, 0.975)
  tbl <- data.frame(term = x_names, estimate = beta_est, conf.low = beta_L, conf.high = beta_U)
  tbl$term <- factor(tbl$term, levels = tbl$term)

  # Sys.setenv(`_R_S3_METHOD_REGISTRATION_NOTE_OVERWRITES_` = "false")
  pCoef <- ggcoef(tbl, ...) + xlab(expression(Posterior ~ ~beta)) + ylab("")
  pCoef
  # dev.off()
}


R/psbcSpeedUp.R
#' psbcSpeedUp
#' @title Function to Fit the Bayesian Cox Lasso Model
#'
#' @description
#' This a speed-up and extended version of the function \code{psbcGL()} in the R package \code{psbcGrouup}
#'
#' @name psbcSpeedUp
#' @docType package
#' @useDynLib psbcSpeedUp
#' @aliases psbcSpeedUp-package
#' @importFrom Rcpp evalCpp
#' @importFrom xml2 as_xml_document write_xml
#' @importFrom stats rexp rgamma runif
#' @importFrom utils write.table
#'
#' @param survObj The list containing observed data from \code{n} subjects;
#' \code{t}, \code{di}, \code{x}. See details for more information
#' @param p number of covariates for variable selection
#' @param q number of mandatory covariates
#' @param hyperpar The list containing prior parameter values; among
#' \code{c('groupInd', 'beta.ini', 'eta0', 'kappa0', 'c0', 'r', 'delta',
#' 'lambdaSq', 'sigmaSq', 'tauSq', 's', 'h', 'beta.prop.var',
#' 'beta.clin.var')}. See details for more information
#' @param nIter the number of iterations of the chain
#' @param burnin number of iterations to discard at the start of the chain.
#' Default is 0
#' @param thin thinning MCMC intermediate results to be stored
#' @param rw When setting to "TRUE", the conventional random walk Metropolis
#' Hastings algorithm is used. Otherwise, the mean and the variance of the
#' proposal density is updated using the jumping rule described in
#' Lee et al. (2011)
#' @param outFilePath path to where the output files are to be written
#' @param tmpFolder the path to a temporary folder where intermediate data
#' files are stored (will be erased at the end of the chain). It is specified
#' relative to \code{outFilePath}
#'
#' @details
#' \tabular{ll}{
#' \code{t} \tab a vector of \code{n} times to the event \cr
#' \code{di} \tab a vector of \code{n} censoring indicators for the event time (1=event occurred, 0=censored) \cr
#' \code{x} \tab covariate matrix, \code{n} observations by \code{p} variables\cr
#' \code{groupInd} \tab a vector of \code{p} group indicator for each variable\cr
#' \code{beta.ini} \tab the starting values for coefficients \eqn{\beta}\cr
#' \code{eta0} \tab scale parameter of gamma process prior for the cumulative baseline hazard, \eqn{\eta_0 > 0}\cr
#' \code{kappa0} \tab shape parameter of gamma process prior for the cumulative baseline hazard, \eqn{\kappa_0 > 0}\cr
#' \code{c0} \tab the confidence parameter of gamma process prior for the cumulative baseline hazard, \eqn{c_0 > 0}\cr
#' \code{r} \tab the shape parameter of the gamma prior for \eqn{\lambda^2}\cr
#' \code{delta} \tab the rate parameter of the gamma prior for \eqn{\lambda^2}\cr
#' \code{lambdaSq} \tab the starting value for \eqn{\lambda^2}\cr
#' \code{sigmaSq} \tab the starting value for \eqn{\sigma^2}\cr
#' \code{tauSq} \tab the starting values for \eqn{\tau^2}\cr
#' \code{s} \tab the set of time partitions for specification of the cumulative baseline hazard function\cr
#' \code{h} \tab the starting values for \eqn{h}\cr
#' \code{beta.prop.var} \tab the variance of the proposal density for \eqn{\beta} in a random walk M-H sampler\cr
#' \code{beta.clin.var} \tab the starting value for the variance of \eqn{\beta}\cr
#' }
#'
#' @return An object of class \code{psbcSpeedUp} is saved as
#' \code{obj_psbcSpeedUp.rda} in the output file, including the following components:
#' \itemize{
#' \item input - a list of all input parameters by the user
#' \item output - a list of the all output estimates:
#' \itemize{
#' \item "\code{beta.p}" - a matrix with MCMC intermediate estimates of the regression coefficients.
#' \item "\code{h.p}" - a matrix with MCMC intermediate estimates of the increments in the cumulative baseline hazard in each interval.
#' \item "\code{tauSq.p}" - a vector MCMC intermediate estimates of the hyperparameter "tauSq".
#' \item "\code{sigmaSq.p}" - a vector MCMC intermediate estimates of the hyperparameter "sigmaSq".
#' \item "\code{lambdaSq.p}" - a vector MCMC intermediate estimates of the hyperparameter "lambdaSq".
#' \item "\code{accept.rate}" - a vector acceptance rates of individual regression coefficients.
#' }
#' \item call - the matched call.
#' }
#'
#'
#' @references Lee KH, Chakraborty S, and Sun J (2011). Bayesian Variable
#' Selection in Semiparametric Proportional Hazards Model for High Dimensional
#' Survival Data. \emph{The International Journal of Biostatistics}, 7(1):1-32.
#' @references Zucknick M, Saadati M, and Benner A (2015). Nonidentical twins:
#' Comparison of frequentist and Bayesian lasso for Cox models.
#' \emph{Biometrical Journal}, 57(6):959–81.
#'
#' @examples
#'
#' # Load the example dataset
#' data("exampleData", package = "psbcSpeedUp")
#' p <- exampleData$p
#' q <- exampleData$q
#' survObj <- exampleData[1:3]
#'
#' # Set hyperparameters
#' mypriorPara <- list(
#'   "groupInd" = 1:p, "eta0" = 0.02, "kappa0" = 1, "c0" = 2, "r" = 10 / 9, 
#'   "delta" = 1e-05, "lambdaSq" = 1, "sigmaSq" = runif(1, 0.1, 10), 
#'   "beta.prop.var" = 1, "beta.clin.var" = 1)
#'
#'\donttest{
#' # run Bayesian Lasso Cox
#' library("psbcSpeedUp")
#' set.seed(123)
#' fitBayesCox <- psbcSpeedUp(survObj,
#'   p = p, q = q, hyperpar = mypriorPara,
#'   nIter = 10, burnin = 0, outFilePath = tempdir()
#' )
#' plot(fitBayesCox, color = "blue")
#'}
#'
#' @export
psbcSpeedUp <- function(survObj = NULL, p = 0, q = 0, hyperpar = list(),
                        nIter = 1, burnin = 0, thin = 1, rw = FALSE,
                        outFilePath, tmpFolder = "tmp/") {
  # Check the survival input object
  if (!is.list(survObj)) {
    stop("Argument 'survObj' must be an input")
  }
  if (sum(names(survObj) %in% c("t", "di", "x")) != 3) {
    stop("List 'survObj' must have three compoents 't', 'di' and 'x'!")
  }
  if (is.data.frame(survObj$x) | is.matrix(survObj$x)) {
    if (is.data.frame(survObj$x)) {
      survObj$x <- data.matrix(survObj$x)
    }
  } else {
    stop("Data 'survObj$x' must be a dataframe or a matrix!")
  }

  if (p + q == 0) {
    p <- ncol(survObj$x)
  } else {
    if (p %% 1 != 0 | p < 1) {
      stop("Argument 'p' must be a positive integer!")
    }
    if (q %% 1 != 0 | q < 0) {
      stop("Argument 'q' must be a positive integer!")
    }
    if (p + q != ncol(survObj$x)) {
      stop("The sum of 'p' and 'q' must equal the number of columns of 'survObj$x'!")
    }
  }
  if (thin %% 1 != 0 | thin < 1) {
    stop("Argument 'thin' must be a positive integer!")
  }

  # Check the directory for the output files
  if (outFilePath == "") {
    stop("Please specify a directory to save all output files!")
  }

  outFilePathLength <- nchar(outFilePath)
  if (substr(outFilePath, outFilePathLength, outFilePathLength) != "/") {
    outFilePath <- paste(outFilePath, "/", sep = "")
  }
  if (!file.exists(outFilePath)) {
    dir.create(outFilePath)
  }

  # Create temporary directory
  tmpFolderLength <- nchar(tmpFolder)
  if (substr(tmpFolder, tmpFolderLength, tmpFolderLength) != "/") {
    tmpFolder <- paste(tmpFolder, "/", sep = "")
  }
  tmpFolder <- paste(outFilePath, tmpFolder, sep = "")
  if (!file.exists(tmpFolder)) {
    dir.create(tmpFolder)
  }

  # check the formula
  cl <- match.call()

  # Write down in a single data file
  write.table(cbind(survObj$t, survObj$di, survObj$x),
    paste(sep = "", tmpFolder, "data.txt"),
    row.names = FALSE, col.names = FALSE
  )
  data <- paste(sep = "", tmpFolder, "data.txt")

  # Check hyperparameters
  if (!is.list(hyperpar)) {
    stop("Argument 'hyperpar' must be a list!")
  }
  if (any(!names(hyperpar) %in%
    c(
      "beta.ini", "lambdaSq", "sigmaSq", "rate", "tauSq",
      "s", "h", "groupInd", "eta0", "kappa0", "c0", "r", "delta",
      "beta.prop.var", "beta.clin.var"
    ))) {
    stop("Hyperparameters must be among c('beta.ini', 'lambdaSq', 'sigmaSq',
         'rate', 'tauSq', 'h', 'groupInd', 'eta0', 'kappa0', 'c0', 'r',
         'delta', 'beta.prop.var', 'beta.clin.var')!")
  }
  if ("groupInd" %in% names(hyperpar)) {
    if (length(hyperpar$groupInd) != p)
      stop("Please specify correct hyperpar$groupInd!")
    groupInd <- hyperpar$groupInd
  } else {
    groupInd <- 1:p
  }
  if ("beta.ini" %in% names(hyperpar)) {
    ini_beta <- hyperpar$beta.ini
  } else {
    ini_beta <- rep(0, p + q)
  }
  if (!"lambdaSq" %in% names(hyperpar)) {
    hyperpar$lambdaSq <- 1
  }
  if (!"sigmaSq" %in% names(hyperpar)) {
    hyperpar$sigmaSq <- runif(1, 0.1, 10)
  }
  if (!"rate" %in% names(hyperpar)) {
    hyperpar$rate <- hyperpar$lambdaSq / 2
  }
  if ("tauSq" %in% names(hyperpar)) {
    ini_tauSq <- hyperpar$tauSq
  } else {
    ini_tauSq <- rexp(length(unique(groupInd))) # , rate = hyperpar$lambdaSq/2)
  }
  s0 <- survObj$t[survObj$di == 1]
  if ("s" %in% names(hyperpar)) {
    if (min(s) > min(s0) | max(s) < max(s0)) {
      stop("Parameter 's' does not cover all event times!")
    }
    if (any(diff(s) < 0)) {
      s <- sort(s)
    }
    if (any(duplicated(s))) {
      s <- unique(s)
    }
  } else {
    s <- sort(unique(s0)) # event times that are actually observed
    s <- c(s, 2 * max(survObj$t) - max(survObj$t[-which.max(survObj$t)]))
  }
  if ("h" %in% names(hyperpar)) {
    ini_h <- hyperpar$h
  } else {
    ini_h <- rgamma(length(s), 1, 1)
  }
  if (!"eta0" %in% names(hyperpar)) {
    hyperpar$eta0 <- round(log(2) / 36, 2)
  }
  if (!"kappa0" %in% names(hyperpar)) {
    hyperpar$kappa0 <- 1
  }
  if (!"c0" %in% names(hyperpar)) {
    hyperpar$c0 <- 2
  }
  if (!"r" %in% names(hyperpar)) {
    hyperpar$r <- 10 / 9
  }
  if (!"delta" %in% names(hyperpar)) {
    hyperpar$delta <- 0.00001
  }
  if ("beta.prop.var" %in% names(hyperpar)) {
    hyperpar$beta_prop_var <- hyperpar$beta.prop.var
  } else {
    hyperpar$beta_prop_var <- 1
  }
  if ("beta.clin.var" %in% names(hyperpar)) {
    hyperpar$beta_clin_var <- hyperpar$beta.clin.var
  } else {
    hyperpar$beta_clin_var <- 1
  }

  hyperParFile <- paste(sep = "", tmpFolder, "hyperpar.xml")

  ## Create the return object
  ret <- list(input = list(), output = list(), call = cl)
  class(ret) <- "psbcSpeedUp"

  # Copy the inputs
  ret$input["p"] <- p
  ret$input["q"] <- q
  ret$input["nIter"] <- nIter
  ret$input["burnin"] <- burnin
  ret$input["thin"] <- thin
  ret$input["rw"] <- rw
  ret$input["outFilePath"] <- outFilePath
  ret$input["tmpFolder"] <- tmpFolder
  ret$input["hyperpar"] <- list(hyperpar)

  hyperpar$s <- hyperpar$beta.ini <- hyperpar$tauSq <- hyperpar$h <- hyperpar$groupInd <-
    hyperpar$beta.prop.var <- hyperpar$beta.clin.var <- NULL

  ## Set up the XML file for hyperparameters
  xml <- xml2::as_xml_document(
    list(hyperparameters = list(
      lapply(hyperpar, function(x) list(format(x, scientific = FALSE))) # every element in the list should be a list
    ))
  )
  xml2::write_xml(xml, file = hyperParFile)

  # Run Bayesian Cox model
  nChains <- 1
  ret$output <- psbcSpeedUp_internal(
    data, p, q, hyperParFile, outFilePath,
    ini_beta, ini_tauSq, ini_h, groupInd, # hyperparameters which are vectors
    nIter, nChains, thin, rw
  )
  
  ret$output$accept.rate <- as.vector(ret$output$accept.rate) / nIter

  if (is.null(colnames(survObj$x))) {
    colnames(ret$output$beta.p) <- paste0("x", 1:ncol(survObj$x))
  } else {
    colnames(ret$output$beta.p) <- colnames(survObj$x)
  }

  ## Save fitted object
  obj_psbc <- list(input = ret$input, output = ret$output)
  save(obj_psbc, file = paste(sep = "", outFilePath, "obj_psbc.rda"))

  if (outFilePath != tmpFolder) {
    unlink(tmpFolder, recursive = TRUE)
  }

  return(ret)
}
